### ISSUE TITLE
Holiday Stations Missing

### ISSUE THREAD

**Issue by lebeno** [Soma-FM-Kodi-Add-On/plugin.audio.somafm#8]
The pluging does not list the SomaFM holiday stations:
- Department Store Christmas
- Jolly Ol' Soul
- Xmas in Frisko
- Christmas Rocks!
- Christmas Lounge

Issue on the old repo: https://github.com/Oderik/xbmc-somafm/issues/11

**Comment by evinr** [comment]
Hey there @lebeno I am not able to reproduce this issue, see attached screenshots. Can you confirm what version of the addon is and what device you are testing on?

![Screenshot from 2021-12-13 08-47-24](https://user-images.githubusercontent.com/1620062/145854079-2f06dcc8-82f1-48ef-a216-f4e8757c7e1f.png)
![Screenshot from 2021-12-13 08-47-54](https://user-images.githubusercontent.com/1620062/145854083-b5c1a53c-7a0a-4562-a589-d3bfe4e3a2a7.png)
![Screenshot from 2021-12-13 08-48-05](https://user-images.githubusercontent.com/1620062/145854088-7ffef00f-793f-4fe1-875d-b7d364ff32c6.png)
?

**Comment by lebeno** [comment]
Addon version 1.1.0 ("Oderik and PanderMusubi")
LibreElec 9.2.8 on RPI4

I suspect the SomaFM station list on my KODI was downloaded/created when I installed/first used the plugin and at that time, the "holiday stations" were not on the list. Probably those stations are only visible on the list around December and the pluging maybe does not update the list regularly?

**Comment by evinr** [comment]
Thanks info! I will keep this in mind as things get updated.

And just to verify, did you try going into the configure tab for Soma FM and clearing the cache explicitly?

**Comment by lebeno** [comment]
I cleared the cache again and I can also confirm that on that settings page it is set for a daily refresh of the channels, but still no holiday channels in the list.

**Comment by evinr** [comment]
Any chance you can manually delete the channels.xml file to verify the clearCache function is not working as expected? https://github.com/Soma-FM-Kodi-Add-On/plugin.audio.somafm/blob/master/resources/lib/addon.py#L212

Mine was located at ~/.kodi/userdata/addon_data/plugin.audio.somafm/channels.xml

**Comment by lebeno** [comment]
Just checked and my channel.xml had a datetime of about 11 hours ago. I did not re-check in the plugin but after deleting the file the plugin did list the holidays channels and also some channels I did not see before (n5MD Radio, Synphaea Radio).

**Comment by lebeno** [comment]
Before deleting the channels.xml I took a copy and opening that file it looks like that file was already up to date with the new channels. Maybe the clearing of the cache did not take into effect immediately or I did not refresh/restart the plugin.

Still, strange that a clear cache was needed to load the new channels.

**Comment by evinr** [comment]
Nice! I am guessing there is an exception that is being thrown here: https://github.com/Soma-FM-Kodi-Add-On/plugin.audio.somafm/blob/master/resources/lib/addon.py#L76

Feel free to add some logging to shine some more light on it. I do not have plans to maintain a Python2 version, but I am assuming this issue affects both versions. The most recent version has the logging, but I think the condition happens when the channels.xml file is updated and the folders are not created. With channels.xml deleted it forces a the desired conditions.

The clearCache is removing the whole directory, which would be great to verify. Would you be able to clear the cache through the settings and then ls the directory to verify the contents?

Perhaps after the holidays when the channels turn off we will have another opportunity to see how it behaves .

**Comment by void-spark** [comment]
https://github.com/Soma-FM-Kodi-Add-On/plugin.audio.somafm/blob/39be4daff4d27120dba4a9b1d5f3062cd4cbd614/resources/lib/addon.py#L64-L68

https://docs.python.org/3/library/os.path.html :
os.path.getmtime(path)
    Return the time of last modification of path. The return value is a floating point number giving the number of **seconds** since the epoch (see the time module). Raise OSError if the file does not exist or is inaccessible.

https://docs.python.org/3/library/time.html :
time.time() → float
    Return the time in **seconds** since the epoch as a floating point number. ...

Yet cache_ttl_in_ms() returns a milliseconds value.
So the cache lasts 1000 times too long :)

Example logged values:
os.path.getmtime: 1639590036.0
time.time: 1639761293.1423855 (difference is 171257, for about 2 days)
cache_ttl_in_ms: 86400000

**Comment by evinr** [comment]
@void-spark what do you think the best solution is? Changing cache_ttl_in_ms() to cache_ttl_in_seconds() and updating the logic?

Also, the cache lasting 5+ years is really a testament to how long Soma FM will be around! 🤣 Are we sure this is a bug and not a feature? j/k

**Comment by void-spark** [comment]
Yeah, just work with seconds here is probably the best option, I can do a PR later today if you like :)

Gotta plan long term :) It's a bit hard to test though, I'm not sure I have devices that manage to run 5 years :D

**Comment by void-spark** [comment]
Ok, see: https://github.com/Soma-FM-Kodi-Add-On/plugin.audio.somafm/pull/13
It should be tested still, ran out of time for that :)
And reviewed of course, but there's really not much to it :)

**Comment by evinr** [comment]
Tested for run-ability and merged. Nice find and great fix! 

After the holiday season ends and the additional stations go on vacation we can close this issue after verification of the cache being cleared before the next holiday season. 📻

**Comment by void-spark** [comment]
Hey, old open tab, did anyone ever check this?  I think we've had one or two seasons since :)
